### PR TITLE
[Core] Sliding window for block manager v2

### LINKED ISSUE #3665: [Misc]: Implement SlidingWindowBlockTable in BlockManagerV2
Recently, we refactored the block manager subsystem to improve testability by separating concerns of each layer. See https://github.com/vllm-project/vllm/pull/3492 for more information.

The V2 implementation does not yet have sliding window support. This issue tracks adding sliding window to the V2 block table so that we can support models that use this feature.

My initial take on the design is to implement a `SlidingWindowBlockTable` that composes within it a `BlockTable`. The `SlidingWindowBlockTable` will then drop blocks that are outside of the context window (potentially mapping them to a devnull block). This will preserve the semantics of the v1 block manager sliding window while fitting into the new design.

### LINKED ISSUE #4057: [Feature][Chunked prefill]: Make sliding window work
### 🚀 The feature, motivation and pitch

Currently, both prefix caching and chunked prefill doesn't work with sliding window attention because the block tables are not updated properly. Both uses `context_attention_forward` (https://github.com/vllm-project/vllm/blob/5c2e66e4871917c5d59cc4a8b89ef53e690e9bd9/vllm/attention/ops/prefix_prefill.py#L621), which I am not sure if sliding window is supported. We should fix this if we want to enable chunked prefill/prefix caching for models that use sliding window attn (e.g., mistral 7B). 

### Alternatives

N/A

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mmoskal**
This implements sliding window in v2 block manager.

First commit comes from #3967 by @ruthe98, but the actual change was somewhat more complex including the concept of a null block.

It passes correctness tests with starcoder3b (the smallest model with sliding window I could find). The test does a bunch of assignments "x1 = 10; x2 = 33; ..." and then asks for value of one of them (which is outside the sliding window). If we tell it upfront which we are going to be looking for, then it answers correctly.

When using chunked prefill all the blocks for prompt are allocated immediately, while we could only allocate enough blocks for the chunk, and free any blocks that are no longer needed. After processing the prompt however, it does free the beginning of prompt at the first generation step.

This can be fixed later. The main problem with fixing this, is that if we're generating more than one sequence, they are all forked in BlockSpaceManagerV2.allocate(), but they really should only be forked after the prompt is fully computed. (see [aborted attempt at fixing this](https://github.com/vllm-project/vllm/pull/4545/commits/9661776b755d7dfce5946bd8830d37a4a30b55e7))

CC @cadedaniel @ruthe98

FIX #3665
FIX #4057 

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by cadedaniel** [comment]
* FYI prefix caching + sliding window doesn't work with block manager v1 yet. so it's ok to prioritize that deprioritize that for this PR https://github.com/vllm-project/vllm/blob/b8afa8b95a4eee008a9b72440620113e5bfbe962/vllm/core/block_manager_v1.py#L218-L220

* personally I think it's OK to allocate entire prompt len for chunked prefill. it's a compute optimization, not a memory capacity optimization, after all

* for testing, we can copy this test except run a sliding window model. it's important to go over the sliding window boundary (although in general the sliding window test coverage is pretty poor). I think one can mock the sliding window size to be much smaller for test convenience, FYI. https://github.com/vllm-project/vllm/blob/b8afa8b95a4eee008a9b72440620113e5bfbe962/tests/core/block/e2e/test_correctness.py#L28

**Comment by mmoskal** [comment]
@cadedaniel @rkooo567 @simon-mo this should be ready for review

**Comment by cadedaniel** [comment]
WTAL on Monday

**Comment by rkooo567** [comment]
QQ: is this the last feature that's needed before enabling block manager v2?

**Comment by cadedaniel** [comment]
See https://github.com/vllm-project/vllm/issues/4537

**Comment by rkooo567** [comment]
Hmm maybe I can help getting cpu swapping done

**Comment by mmoskal** [comment]
@cadedaniel let me know if you need any more info from my side!

File: tests/core/block/e2e/test_correctness_sliding_window.py
Comment (cadedaniel) [inline tests/core/block/e2e/test_correctness_sliding_window.py:109]: I'm afraid these tests won't catch issues with the block mapping. E.g. I expect error to accumulate over many tokens before we see a significant divergence in attention scores. 10/4096 tokens is not very much, same for 128/4096 although it's better.

WDYT? Is my intuition right? Should we test with larger generation size? Another option is to patch `sliding_window` to be smaller (e.g. two blocks) so the impact of any error is larger. If we go with patching `sliding_window` we could even use one of the 68m models for a faster test.

File: tests/core/block/e2e/test_correctness_sliding_window.py
Comment (cadedaniel) [inline tests/core/block/e2e/test_correctness_sliding_window.py:123]: nit: can we move this and [`get_token_ids_from_llm_generator`](https://github.com/vllm-project/vllm/blob/main/tests/core/block/e2e/test_correctness.py#L449) to the [conftest](https://github.com/vllm-project/vllm/blob/main/tests/core/block/e2e/conftest.py)?

**Review by cadedaniel — COMMENTED** [review]
Did a pass over the correctness test. I'll do a pass over the implementation now.

File: vllm/worker/cache_engine.py
Comment (cadedaniel) [inline vllm/worker/cache_engine.py:53]: Can we get rid of this and the attn_backend interface changes if we zero-out the entire cache from the start?

File: vllm/core/block/block_table.py
Comment (cadedaniel) [inline vllm/core/block/block_table.py:124]: nit: I suggest moving this to its own class `SlidingWindowBlockTable` so the complexity is separated

File: vllm/core/block/block_table.py
Comment (cadedaniel) [inline vllm/core/block/block_table.py:95]: Can you explain why we need this? is it sliding-window-specific or sliding-window+chunked-prefill specific?

File: vllm/core/block/block_table.py
Comment (cadedaniel) [inline vllm/core/block/block_table.py:116]: can we have some comments on this code? (what is this branch for?)

Comment (cadedaniel) [reply vllm/core/block/block_table.py:116]: can we add a unit test for this? see others here https://github.com/vllm-project/vllm/blob/main/tests/core/block/test_block_table.py

File: vllm/core/block/block_table.py
Comment (cadedaniel) [inline vllm/core/block/block_table.py:122]: during forking, is there some risk the python object equality will not hold? should we check the block number (or implement better equality)

File: vllm/core/block/cpu_gpu_block_allocator.py
Comment (cadedaniel) [inline vllm/core/block/cpu_gpu_block_allocator.py:116]: let's add a good docstring here for this new concept

File: vllm/core/block/cpu_gpu_block_allocator.py
Comment (cadedaniel) [inline vllm/core/block/cpu_gpu_block_allocator.py:123]: this is hacky, can we do this in a cleaner way?

Comment (cadedaniel) [reply vllm/core/block/cpu_gpu_block_allocator.py:123]: e.g. a new special block implementation that defers to the internal one except for `append_token_ids`

File: vllm/core/block/cpu_gpu_block_allocator.py
Comment (cadedaniel) [inline vllm/core/block/cpu_gpu_block_allocator.py:166]: Add comment for this new branch ?

Comment (cadedaniel) [reply vllm/core/block/cpu_gpu_block_allocator.py:166]: Hm, do we need to have this? It seems a little weird to have `1` block used even when the system has no requests.

**Review by cadedaniel — COMMENTED** [review]
Pretty nice! some questions / suggestions. mostly I hope we can avoid the `attn_backend` changes.

**Comment by rkooo567** [comment]
oh didn't know this also covers chunked prefill. I will make sure to finish review it by tmrw!

File: vllm/core/block/block_table.py
Comment (rkooo567) [inline vllm/core/block/block_table.py:40]: can you update docstring? Also mention None means sliding window is not used

File: vllm/core/block/block_table.py
Comment (rkooo567) [inline vllm/core/block/block_table.py:118]: is it block idx? Consider renaming end_block_idx?

File: vllm/core/block/block_table.py
Comment (rkooo567) [inline vllm/core/block/block_table.py:95]: nit: num_computed_tokens?

File: vllm/core/block/block_table.py
Comment (rkooo567) [inline vllm/core/block/block_table.py:117]: Can you add comment for this branch? I assume it means chunked prefill is not used?

File: vllm/core/block/cpu_gpu_block_allocator.py
Comment (rkooo567) [inline vllm/core/block/cpu_gpu_block_allocator.py:118]: Not super familiar with interface, but if we use allocate_immutable, isn't this supposed to guarantee the hack below?

File: vllm/core/block/cpu_gpu_block_allocator.py
Comment (rkooo567) [inline vllm/core/block/cpu_gpu_block_allocator.py:116]:
```suggestion
    def allocate_or_get_null_block(self) -> Block:
```

File: vllm/core/block_manager_v2.py
Comment (rkooo567) [inline vllm/core/block_manager_v2.py:74]: why do we need + 2 here?

**Review by rkooo567 — COMMENTED** [review]
~~Hmm I am surprising this test passes without commenting out~~ https://github.com/vllm-project/vllm/blob/cc466a32903d53d0ceca459b766d74ad668c8f87/vllm/worker/model_runner.py#L272. ~~Do you know why? (is there anything missing from test?)~~

Ah, nvm. Guess it doesn't support prefix caching still. Can you just double check the test fails without this changes?

File: vllm/core/block/block_table.py
Comment (mmoskal) [reply vllm/core/block/block_table.py:95]: well, we have `num_full_slots`; it's really the slots (or rather the data in them) that's being computed, not the tokens

Comment (rkooo567) [reply vllm/core/block/block_table.py:95]: hmm but you are passing num_computed_tokens to this API? What's the definition of slots here?

Comment (mmoskal) [reply vllm/core/block/block_table.py:95]: I would say it's the same thing, it's just strange that they are named "computed tokens" at the call site - it's the KV entries being computed, not tokens (tokens are known beforehand)

Comment (mmoskal) [reply vllm/core/block/block_table.py:95]: but happy to change if you feel strongly about it!

Comment (rkooo567) [reply vllm/core/block/block_table.py:95]: oh I was the one who decided that name lol. I think it was because 1 token == 1 kv. I am fine with keeping it slots (or renaming other name). But feel like it is better being consistent

File: vllm/core/block/block_table.py
Comment (mmoskal) [reply vllm/core/block/block_table.py]: When I tried that, I had to repeat the whole body of the fork() method; now this is not big deal, but if someone adds a field to BlockTable, they may forget to update the other fork in SlidingWindowBlockTable. I'm not sure it's worth it. WDYT?

File: vllm/core/block/block_table.py
Comment (mmoskal) [reply vllm/core/block/block_table.py:129]: changed to `isinstance(last_block, NullBlock)`

File: vllm/core/block/cpu_gpu_block_allocator.py
Comment (mmoskal) [reply vllm/core/block/cpu_gpu_block_allocator.py]: It doesn't seem to prevent modifications, at least for the `NaiveBlocks` - we now use a `NullBlock` wrapper anyways. `allocate_immutable()` makes the block participate in hash tables of valid prefixes. I'm not sure we want the null block there (and even then, it's probably not very useful), so I kept `allocate_mutable()`

File: vllm/core/block_manager_v2.py
Comment (mmoskal) [reply vllm/core/block_manager_v2.py:74]: added comment

File: vllm/worker/cache_engine.py
Comment (cadedaniel) [reply vllm/worker/cache_engine.py]: asking internally for a vibe check

File: vllm/core/block/cpu_gpu_block_allocator.py
Comment (cadedaniel) [inline vllm/core/block/cpu_gpu_block_allocator.py:244]: love the clean design !

File: tests/core/block/e2e/test_correctness_sliding_window.py
Comment (mmoskal) [reply tests/core/block/e2e/test_correctness_sliding_window.py:94]: There was an issue with the block tables passed to single token decode, which was causing different output with 1024 tokens. I have now fixed that and bumped test size to 1024.

However, it's still slightly incorrect because the decode kernel does not support sliding window natively - the way it works now it just takes all the blocks passed in (up to seq_len). With v1 manager, the sliding window uses blocks in a "ring buffer" fashion, so this is not a problem. With the new block manager we need potentially to start attention computation in the middle of a block, otherwise we pay attention to a few tokens too many. It doesn't seem to affect this test though.

I have [started fixing](https://github.com/vllm-project/vllm/pull/4545/commits/609a9cebe2ddcfe7c9390c567f50b552fb22a6f3) the decode kernel, but I think that should be a separate PR.

Comment (cadedaniel) [reply tests/core/block/e2e/test_correctness_sliding_window.py:94]: I haven't looked at the changes but want to say that yes, this problem is known and we should fix it eventually (awesome if you want to do it). let's get this PR in with good tests for where we're at and future PR can fix the decode kernel.

File: vllm/core/block/block_table.py
Comment (mmoskal) [reply vllm/core/block/block_table.py]: add test

**Comment by mmoskal** [comment]
@cadedaniel @rkooo567 should be good for another review

**Comment by mmoskal** [comment]
@cadedaniel @rkooo567 anything I can help with?

**Comment by rkooo567** [comment]
I will have the final look today!

**Comment by rkooo567** [comment]
sorry it's been slipped! I will take a look at it by tmrw!

File: vllm/core/block_manager_v2.py
Comment (rkooo567) [inline vllm/core/block_manager_v2.py:72]:
```suggestion
        self.max_block_sliding_window = None
```
?

File: tests/core/block/e2e/test_correctness_sliding_window.py
Comment (rkooo567) [inline tests/core/block/e2e/test_correctness_sliding_window.py:155]: is this the higest we can get? Or more of arbitrarty value?

File: tests/core/block/e2e/test_correctness_sliding_window.py
Comment (rkooo567) [inline tests/core/block/e2e/test_correctness_sliding_window.py:72]: Hmm but this compares block manager v1 vs v2? Doesn't this mean everything else is equivalent (i.e., kernels and things like that).

cc @cadedaniel do you think this is expected outputs are different?

**Review by rkooo567 — COMMENTED** [review]
Mostly looks good. One question for the test (feel a little sketchy why baseline vs test has different output). I will approve after addressing this.

File: tests/core/block/e2e/test_correctness_sliding_window.py
Comment (mmoskal) [reply tests/core/block/e2e/test_correctness_sliding_window.py:72]: Unfortunately, it's not the same kernel arguments. With v2 manager we're passing up 15 kv entries more to the paged attention kernel. With v1, the kv entries "wrap around" in the blocks, but they do not with v2 (needed for prefix caching etc.), see description in #4768

File: vllm/core/block_manager_v2.py
Comment (mmoskal) [reply vllm/core/block_manager_v2.py]: done

**Review by rkooo567 — APPROVED** [review]
LGTM. @cadedaniel can you have a final look?

File: tests/core/block/e2e/test_correctness_sliding_window.py
Comment (rkooo567) [inline tests/core/block/e2e/test_correctness_sliding_window.py:122]: any good way to assert the return prompt token len > 4k?

File: tests/core/block/e2e/test_correctness_sliding_window.py
Comment (cadedaniel) [reply tests/core/block/e2e/test_correctness_sliding_window.py:72]: yeah, the reason why this is so hard to test is because the semantics of blocks are changed in V2. E.g. we now have clear distinction between mutable and immutable blocks. so the kernels that previously would overwrite blocks (causing U.D. with copy-on-write in V1) now don't, but the downside is they capture additional context since we don't yet have masking.

The tests in this PR are not really good enough to catch correctness issues with the sliding window block mapping. The error tolerance in this test is very high and the unit test `test_sliding_window` only checks num consumed is correct.

That said, this is an improvement over the previous sliding window tests and I think we can merge and follow up later..

FWIW the way I'd test this is one/both of the following:
* Modify this test to use block_size=1. This avoids the masking issue entirely and we should expect exact equality between v1 and v2 for most prompts.
* Add a stronger unit test for block_manager_v2 or block_allocator that verifies the correct sliding window block mapping

**Review by cadedaniel — APPROVED** [review]
LGTM. thanks.

**Comment by cadedaniel** [comment]
@mmoskal can you check the merge conflict? will merge after

**Comment by mmoskal** [comment]
Unfortunately, after merge the tests stopped working. The problem is that it's also the baseline tests (not using v2 block manager) that are not working. I'm getting the first token of the output correct, and the remaining tokens not complete gibberish but also not correct - so this is a problem with the decode phase or maybe kv cache entry arrangement?

I tried reverting all my changes in model_runner.py and the baseline tests still fail, which suggests it's something in the recent changes.

**Comment by mmoskal** [comment]
OK should work now - I fixed `slot_mapping` computation in `_prepare_model_input`

**Comment by cadedaniel** [comment]
as soon are tests are green let's merge. cc @rkooo567 for next week.

**Comment by mmoskal** [comment]
the failing tests don't look related to what I'm doing; I just tried pushing a random change  to re-run

**Comment by mmoskal** [comment]
@rkooo567 @cadedaniel tests are green, please merge!

**Comment by rkooo567** [comment]
Thanks for the contribution! Should we next resume the paged attn PR?

**Comment by mmoskal** [comment]
Thank you for merging! I probably won't have time to work on the paged attn kernel PR in the next few weeks :/ The thing is, with this PR the paged attention is almost correct, it just pays attention to a few tokens too many.

**Pull request merged**
